storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -4010,6 +3994,32 @@ dberr_t fil_space_t::garbage_collect(bool shutdown)
     return err;
   }
 
+  /* Remove cached undo log segments in the rollback segment
+  header page exist in system tablespace */
+  mtr.start();
+  mtr.x_lock_space(fil_system.sys_space);
+  for (ulint rseg_id= 0; rseg_id < TRX_SYS_N_RSEGS; rseg_id++)
+  {
+    trx_rseg_t *rseg= &trx_sys.rseg_array[rseg_id];
+    if (rseg->space->id == 0)
```

Comment:
We don’t need `rseg_id` for anything here. I’d use a C++11 style enumeration:
```c++
  for (trx_rseg_t &rseg : trx_sys.rseg_array)
  {
    if (rseg.space == fil_system.sys_space)
```
We can also avoid dereferencing `trx_rseg_t::space` by comparing it to `fil_system.sys_space`, which we already have accessed in this function.

I was wondering if it is safe to access 128 pages in a single mini-transaction. It seems to be safe, because the `trx_rseg_t::reinit()` call is not accessing any further pages in the buffer pool. The minimum size of a buffer pool should be larger than 128 pages.

But do we actually need to access anything than the first rollback segment? Isn’t this code only being executed when multiple undo tablespaces exist? In that case, shouldn’t the only rollback segment 0 reside in the system tablespace?

Reply:
We can shrink the system tablespace even though we don't have external undo tablespace. (innodb.undo_leak test case does it). We're resetting TRX_RSEG_N_SLOTS in all rollback segments. Treat the cached undo segment
as unused one and free it. This is done only after ensuring that no undo logs exist to purge.